fendermint/app/src/app.rs

```diff
@@ -280,15 +283,45 @@ where
     /// Take the execution state, update it, put it back, return the output.
     async fn modify_exec_state<T, F, R>(&self, f: F) -> Result<T>
     where
-        F: FnOnce((CheckpointPool, FvmExecState<SS>)) -> R,
-        R: Future<Output = Result<((CheckpointPool, FvmExecState<SS>), T)>>,
+        F: FnOnce((CheckpointPool, Arc<P>, FvmExecState<SS>)) -> R,
+        R: Future<Output = Result<((CheckpointPool, Arc<P>, FvmExecState<SS>), T)>>,
     {
         let state = self.take_exec_state();
-        let ((_pool, state), ret) = f((self.resolve_pool.clone(), state)).await?;
+        let ((_pool, _provider, state), ret) = f((
+            self.resolve_pool.clone(),
+            self.parent_finality_provider.clone(),
+            state,
+        ))
+        .await?;
         self.put_exec_state(state);
         Ok(ret)
     }
 
+    /// Get a read only fvm execution state. This is useful to perform query commands targeting
+    /// the latest state.
+    pub fn new_read_only_exec_state(
+        &self,
+    ) -> Result<Option<FvmExecState<ReadOnlyBlockstore<Arc<SS>>>>> {
+        let maybe_app_state = self.get_committed_state()?;
+
+        Ok(if let Some(app_state) = maybe_app_state {
+            let block_height = app_state.block_height;
```

Comment:
I think this should also return `None` if the `block_height` is `None`. You won't be able to execute queries in that state. This is checked in the `query` method, because the state exists (put there by `init_committed_state` for the `info` method for example, but we won't be able to create an `exec_state` until after `genesis`.

Reply:
oh, you mean `block_height == 0`? I will add this check.